src/Compilers/Core/Portable/Symbols/TypedConstant.cs

```diff
@@ -117,7 +117,7 @@ public ImmutableArray<TypedConstant> Values
 
                 if (this.IsNull)
```

Comment:
i'd actually prefer we just doc this as the expected behavior of `.Values`.  Given that `null` is a legal value to pass for an array, it would be good to be able to distinguish an empty set of values or a null array.

Reply:
Currently it returns `default` - Which isn't `null` for this type. It returns an uninitialised array that throws an exception when you try to iterate through it.

Reply:
It should continue to return `default`, but the method should be doc'ed to indicate this is a legal value that you get if `null` was passed as the array value.  The docs can also mention that `<see name="IsNull"/>` can be used to easily check for this.

Reply:
> Right. I'm saying that i think this is good.

Oh right. It just leads to more verbose code when trying to pull out values. For instance, if I want to collect values but not throw an exception, I'll have to do something like this:
```diff
- var values = attribute.ConstructorArguments.Select(x => x.Values);
+ var values = attribute.ConstructorArguments.Select(x => x.Values.IsDefaultOrEmpty ? x.Values : ImmutableArray<TypedConstant>.Empty);
```

But I guess it is still good to be able to distinguish null or not null.

However, this still shouldn't be throwing a NullReferenceException in analyzers. It's deceiving and doesn't show the true error.

Reply:
I dont' think ti's deceiving if we doc that you get a `default` immutable array.  That's just the behavior you get if you iterate in that fashion.  You coudl report a request to the runtime that iterating should throw such an exception.

Reply:
Ah okay. I'm with you. I just found the error super vague and unhelpful. As the exception is just that one line, and there's no stacktrace, so it was really difficult to find the root cause!

Reply:
If we want to distinguish between null and not null, couldn't this `Values` property just be nullable and be set to null?

Currently it's not nullable, but does throw a NullReferenceException, which isn't a great experience when debugging or encountering errors.

Reply:
> If we want to distinguish between null and not null, couldn't this Values property just be nullable and be set to null?

ImmutableArray already has a way to repesent null (which is what is used here).  So it would be superfluous.  It would also be a binary breaking change to chagen the return type of this property.

> Currently it's not nullable, but does throw a NullReferenceException, which isn't a great experience when debugging or encountering errors.

This is the nature of ImmutableArrays.  Someone could write an analyzer to allow these to be annotated and reported on at the consumption side (like NRTs).